<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Example usage &mdash; labtools  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
      <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Changelog" href="changelog.html" />
    <link rel="prev" title="labtools" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            labtools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Example usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sequence-design-tools">Sequence Design Tools</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#imports">Imports</a></li>
<li class="toctree-l3"><a class="reference internal" href="#shuffle-a-sequence">Shuffle a sequence</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sequencing-analysis-tools">Sequencing Analysis Tools</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sort-processing-example">Sort Processing Example</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#what-you-need">What you need</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-you-get-out">What you get out</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id1">Imports</a></li>
<li class="toctree-l3"><a class="reference internal" href="#initialize-a-sort">Initialize a sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-activity-values">Get activity values</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#locate-any-tiles-not-just-perfect-matches">Locate any tiles, not just perfect matches</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#get-data-for-reads-that-include-barcodes-and-tiles">Get data for reads that include barcodes AND tiles</a></li>
<li class="toctree-l3"><a class="reference internal" href="#support-for-custom-anchor-sequences">Support for custom anchor sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#use-the-sequence-directly-preceding-the-ad-tile-as-an-anchor-sequence-additional-characters-between-ad-tile-and-preceding-anchor-seqeunce-will-not-work">Use the sequence directly preceding the AD/tile as an anchor sequence. Additional characters between AD/tile and preceding anchor seqeunce will not work</a></li>
<li class="toctree-l4"><a class="reference internal" href="#specify-alternate-values-in-a-dictionary-and-pass-that-dictionary-to-sort-process">Specify alternate values in a dictionary and pass that dictionary to Sort.process()</a></li>
<li class="toctree-l4"><a class="reference internal" href="#example-output-for-pull-ad">Example output for pull_AD</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#counting-tiles-in-a-fastq-file">Counting tiles in a fastq file</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#use-barcoded-true-to-count-tile-and-barcode-pairs">Use barcoded = True  to count tile and barcode pairs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#use-the-design-file-flag-to-look-for-specific-tiles">Use the design_file flag to look for specific tiles</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="conduct.html">Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference internal" href="autoapi/index.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">labtools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Example usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/example.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="example-usage">
<h1>Example usage<a class="headerlink" href="#example-usage" title="Permalink to this heading"></a></h1>
<p>To use <code class="docutils literal notranslate"><span class="pre">labtools</span></code> in a project:</p>
<section id="sequence-design-tools">
<h2>Sequence Design Tools<a class="headerlink" href="#sequence-design-tools" title="Permalink to this heading"></a></h2>
<section id="imports">
<h3>Imports<a class="headerlink" href="#imports" title="Permalink to this heading"></a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">labtools</span> <span class="kn">import</span> <span class="n">shuffle</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="shuffle-a-sequence">
<h3>Shuffle a sequence<a class="headerlink" href="#shuffle-a-sequence" title="Permalink to this heading"></a></h3>
<p>Create shuffles of the alphabet:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">shuffles_list</span><span class="p">,</span> <span class="n">names_list</span> <span class="o">=</span> <span class="n">shuffle</span><span class="o">.</span><span class="n">windowed_shuffle</span><span class="p">(</span><span class="s2">&quot;ABCDEFGHIJKLMNOPQRSTUVWYXZ&quot;</span><span class="p">)</span>
<span class="n">shuffles_list</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;ceadbFGHIJKLMNOPQRSTUVWYXZ&#39;,
 &#39;AedbfcGHIJKLMNOPQRSTUVWYXZ&#39;,
 &#39;ABgdcfeHIJKLMNOPQRSTUVWYXZ&#39;,
 &#39;ABCeghdfIJKLMNOPQRSTUVWYXZ&#39;,
 &#39;ABCDiegfhJKLMNOPQRSTUVWYXZ&#39;,
 &#39;ABCDEhifgjKLMNOPQRSTUVWYXZ&#39;,
 &#39;ABCDEFkigjhLMNOPQRSTUVWYXZ&#39;,
 &#39;ABCDEFGkjilhMNOPQRSTUVWYXZ&#39;,
 &#39;ABCDEFGHjkilmNOPQRSTUVWYXZ&#39;,
 &#39;ABCDEFGHInmljkOPQRSTUVWYXZ&#39;,
 &#39;ABCDEFGHIJmnolkPQRSTUVWYXZ&#39;,
 &#39;ABCDEFGHIJKnlpomQRSTUVWYXZ&#39;,
 &#39;ABCDEFGHIJKLnqpmoRSTUVWYXZ&#39;,
 &#39;ABCDEFGHIJKLMnropqSTUVWYXZ&#39;,
 &#39;ABCDEFGHIJKLMNsrqopTUVWYXZ&#39;,
 &#39;ABCDEFGHIJKLMNOtprqsUVWYXZ&#39;,
 &#39;ABCDEFGHIJKLMNOPstrquVWYXZ&#39;,
 &#39;ABCDEFGHIJKLMNOPQvtsruWYXZ&#39;,
 &#39;ABCDEFGHIJKLMNOPQRwsutvYXZ&#39;,
 &#39;ABCDEFGHIJKLMNOPQRSuywvtXZ&#39;,
 &#39;ABCDEFGHIJKLMNOPQRSTyxwuvZ&#39;,
 &#39;ABCDEFGHIJKLMNOPQRSTUzvxwy&#39;]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="sequencing-analysis-tools">
<h2>Sequencing Analysis Tools<a class="headerlink" href="#sequencing-analysis-tools" title="Permalink to this heading"></a></h2>
<section id="sort-processing-example">
<h3>Sort Processing Example<a class="headerlink" href="#sort-processing-example" title="Permalink to this heading"></a></h3>
<section id="what-you-need">
<h4>What you need<a class="headerlink" href="#what-you-need" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>fastq file for each sample in the sort</p></li>
<li><p>bin counts (cells per bin) and bin values (mean or median fluorescence of the bin)</p></li>
<li><p><strong>if you want perfect matches</strong>: a csv with one column headered as “ArrayDNA” which contains your 120 bp AD sequences as DNA</p></li>
<li><p>the sequence directly preceding your AD sequence in your reads (anchor sequence)</p></li>
<li><p>if you have barcodes, you need the sequence directly preceding them and anteceding them (anchor sequences)</p></li>
</ul>
</section>
<section id="what-you-get-out">
<h4>What you get out<a class="headerlink" href="#what-you-get-out" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>a dataframe with your AD tiles as indices, normalized scores for tile at each bin, and the tile Activity value</p></li>
</ul>
<p><strong>NOTE: the read cutoff is 10 reads total summed across all bins</strong> This means that if a tile is not found at least 10 times combined across any of the fastq files, it will not be analysed.</p>
</section>
</section>
<section id="id1">
<h3>Imports<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">labtools.adtools</span> <span class="kn">import</span> <span class="n">sort</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="initialize-a-sort">
<h3>Initialize a sort<a class="headerlink" href="#initialize-a-sort" title="Permalink to this heading"></a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">my_sort</span> <span class="o">=</span> <span class="n">sort</span><span class="o">.</span><span class="n">Sort</span><span class="p">([</span><span class="s2">&quot;../exampledata/bin1.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;../exampledata/bin2.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;../exampledata/bin3.fastq&quot;</span><span class="p">,</span> 
                <span class="s2">&quot;../exampledata/bin4.fastq&quot;</span><span class="p">],</span> <span class="n">bin_counts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100000</span><span class="p">,</span><span class="mi">100000</span><span class="p">,</span><span class="mi">100000</span><span class="p">,</span><span class="mi">100000</span><span class="p">],</span> 
               <span class="n">bin_values</span> <span class="o">=</span> <span class="p">[</span><span class="mi">61</span><span class="p">,</span><span class="mi">141</span><span class="p">,</span><span class="mi">251</span><span class="p">,</span><span class="mi">1462</span><span class="p">],</span> <span class="n">design_file</span> <span class="o">=</span> <span class="s2">&quot;../exampledata/unique_seqs.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="get-activity-values">
<h3>Get activity values<a class="headerlink" href="#get-activity-values" title="Permalink to this heading"></a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">activities</span><span class="p">,</span> <span class="n">total_reads</span><span class="p">,</span> <span class="n">reads_per_bin</span> <span class="o">=</span> <span class="n">my_sort</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># activity + normalized abundance in each bin</span>
<span class="n">activities</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>Activity</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTAAATTCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT</th>
      <td>0.279148</td>
      <td>0.293312</td>
      <td>0.427540</td>
      <td>0.000000</td>
      <td>165.697515</td>
    </tr>
    <tr>
      <th>GAAGCTTTGTCTTTGTGTGATGATTTGTTGGGTGATCAAGATAGAGAAGAGAATGATAATGATGGTGATTTGAAAGATAAACAACCATGTATTGCTGATTATGCTCATTTGGGTCCAGAA</th>
      <td>0.110487</td>
      <td>0.540278</td>
      <td>0.325749</td>
      <td>0.023487</td>
      <td>199.019328</td>
    </tr>
    <tr>
      <th>GATTTGGCTGAAGATGATGAAGTTATGTGTATGGAAGATGAAGTTCAATCTATTCAACCAAATCATGAAAGACCAGATGATGGTCCAGAATTGGAACATGGTTTGGAGAATGGTGCTAAA</th>
      <td>0.376741</td>
      <td>0.492582</td>
      <td>0.130676</td>
      <td>0.000000</td>
      <td>125.235085</td>
    </tr>
    <tr>
      <th>TCTGAATTGACTTCTACTTTGGGTATTTCTCATAGATTGCCACAATCTTTGACTCCATGTGTTAAGACTGGTTCTTTGCAATCTGGTGGTTTGGTTCAATCTGTTCCATTTGAAGAATTG</th>
      <td>0.362058</td>
      <td>0.279914</td>
      <td>0.358029</td>
      <td>0.000000</td>
      <td>151.418558</td>
    </tr>
    <tr>
      <th>AATCATGAAAGACCATCTTCTGGTCCAGAAGCTGAACATGGTTTGGAGAATGGTGCTAAAGAAATGGCTGATGATAAAGAAGAACAAGAGAAAGATAGAGATAATGAGAATCAAGGTGAA</th>
      <td>0.611774</td>
      <td>0.247237</td>
      <td>0.093699</td>
      <td>0.047290</td>
      <td>164.835151</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>GACCCAACTGAATGGTTTGATTCTGGTGCTCAATTTATCTTGAATGCTCAACAATTGGTTGAAGCTCAATGTTTGGATGATAATTTGACTAGAGAATTGGAATCTAATGATGGTGCTTTG</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>TCTACTGATTCTACTCCAATGTTTGATTATGATAATTTGGAAGATAATTCTAAAGATTGGACTTCTTTGTTTGATAATGATATTCCAGTTACTACTGATGATGTTTCTTTGGCTGATAAA</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>TCTACTGGTCAAGTCTTGTTTGATATTGATGACTTTAGATGGTTGTTGGATCCAGATGATGAACAATTGGGTAAAGAAGCTATCTTGTCTGATCAATTTGGTAAACCAACTCCAGAGAAT</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>GAAGATCCAACTTCTGATTCTGCTATTCAACAATTGTGGAATCAAGGATTCTTGTTTGTTGAATCTTTGTCTTTGTGTGATGATTTGTTGGGTTCTCAAGATAGAGAAGAGAATACTAAT</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>GAAATTGATCAAATTTCTGATCCAGATAAATTGCCAGTTAATTTGGAACCATTTAGATTGGATCAATTGGAATTTACTGGTGATGATACTTCTGGTGCTGGTTTGAAATTTCAATGGGAT</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
  </tbody>
</table>
<p>101 rows × 5 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># the total reads for each sequence</span>
<span class="n">total_reads</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTAAATTCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT    44.0
GAAGCTTTGTCTTTGTGTGATGATTTGTTGGGTGATCAAGATAGAGAAGAGAATGATAATGATGGTGATTTGAAAGATAAACAACCATGTATTGCTGATTATGCTCATTTGGGTCCAGAA    21.0
GATTTGGCTGAAGATGATGAAGTTATGTGTATGGAAGATGAAGTTCAATCTATTCAACCAAATCATGAAAGACCAGATGATGGTCCAGAATTGGAACATGGTTTGGAGAATGGTGCTAAA    49.0
TCTGAATTGACTTCTACTTTGGGTATTTCTCATAGATTGCCACAATCTTTGACTCCATGTGTTAAGACTGGTTCTTTGCAATCTGGTGGTTTGGTTCAATCTGTTCCATTTGAAGAATTG    70.0
AATCATGAAAGACCATCTTCTGGTCCAGAAGCTGAACATGGTTTGGAGAATGGTGCTAAAGAAATGGCTGATGATAAAGAAGAACAAGAGAAAGATAGAGATAATGAGAATCAAGGTGAA    20.0
                                                                                                                            ... 
GACCCAACTGAATGGTTTGATTCTGGTGCTCAATTTATCTTGAATGCTCAACAATTGGTTGAAGCTCAATGTTTGGATGATAATTTGACTAGAGAATTGGAATCTAATGATGGTGCTTTG    23.0
TCTACTGATTCTACTCCAATGTTTGATTATGATAATTTGGAAGATAATTCTAAAGATTGGACTTCTTTGTTTGATAATGATATTCCAGTTACTACTGATGATGTTTCTTTGGCTGATAAA    24.0
TCTACTGGTCAAGTCTTGTTTGATATTGATGACTTTAGATGGTTGTTGGATCCAGATGATGAACAATTGGGTAAAGAAGCTATCTTGTCTGATCAATTTGGTAAACCAACTCCAGAGAAT    12.0
GAAGATCCAACTTCTGATTCTGCTATTCAACAATTGTGGAATCAAGGATTCTTGTTTGTTGAATCTTTGTCTTTGTGTGATGATTTGTTGGGTTCTCAAGATAGAGAAGAGAATACTAAT    11.0
GAAATTGATCAAATTTCTGATCCAGATAAATTGCCAGTTAATTTGGAACCATTTAGATTGGATCAATTGGAATTTACTGGTGATGATACTTCTGGTGCTGGTTTGAAATTTCAATGGGAT    13.0
Length: 101, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># the reads of each tile per bin</span>
<span class="n">reads_per_bin</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTAAATTCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT</th>
      <td>11.0</td>
      <td>13.0</td>
      <td>20.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>GAAGCTTTGTCTTTGTGTGATGATTTGTTGGGTGATCAAGATAGAGAAGAGAATGATAATGATGGTGATTTGAAAGATAAACAACCATGTATTGCTGATTATGCTCATTTGGGTCCAGAA</th>
      <td>2.0</td>
      <td>11.0</td>
      <td>7.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>GATTTGGCTGAAGATGATGAAGTTATGTGTATGGAAGATGAAGTTCAATCTATTCAACCAAATCATGAAAGACCAGATGATGGTCCAGAATTGGAACATGGTTTGGAGAATGGTGCTAAA</th>
      <td>17.0</td>
      <td>25.0</td>
      <td>7.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>TCTGAATTGACTTCTACTTTGGGTATTTCTCATAGATTGCCACAATCTTTGACTCCATGTGTTAAGACTGGTTCTTTGCAATCTGGTGGTTTGGTTCAATCTGTTCCATTTGAAGAATTG</th>
      <td>23.0</td>
      <td>20.0</td>
      <td>27.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>AATCATGAAAGACCATCTTCTGGTCCAGAAGCTGAACATGGTTTGGAGAATGGTGCTAAAGAAATGGCTGATGATAAAGAAGAACAAGAGAAAGATAGAGATAATGAGAATCAAGGTGAA</th>
      <td>11.0</td>
      <td>5.0</td>
      <td>2.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>GACCCAACTGAATGGTTTGATTCTGGTGCTCAATTTATCTTGAATGCTCAACAATTGGTTGAAGCTCAATGTTTGGATGATAATTTGACTAGAGAATTGGAATCTAATGATGGTGCTTTG</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>23.0</td>
    </tr>
    <tr>
      <th>TCTACTGATTCTACTCCAATGTTTGATTATGATAATTTGGAAGATAATTCTAAAGATTGGACTTCTTTGTTTGATAATGATATTCCAGTTACTACTGATGATGTTTCTTTGGCTGATAAA</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>24.0</td>
    </tr>
    <tr>
      <th>TCTACTGGTCAAGTCTTGTTTGATATTGATGACTTTAGATGGTTGTTGGATCCAGATGATGAACAATTGGGTAAAGAAGCTATCTTGTCTGATCAATTTGGTAAACCAACTCCAGAGAAT</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>12.0</td>
    </tr>
    <tr>
      <th>GAAGATCCAACTTCTGATTCTGCTATTCAACAATTGTGGAATCAAGGATTCTTGTTTGTTGAATCTTTGTCTTTGTGTGATGATTTGTTGGGTTCTCAAGATAGAGAAGAGAATACTAAT</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>11.0</td>
    </tr>
    <tr>
      <th>GAAATTGATCAAATTTCTGATCCAGATAAATTGCCAGTTAATTTGGAACCATTTAGATTGGATCAATTGGAATTTACTGGTGATGATACTTCTGGTGCTGGTTTGAAATTTCAATGGGAT</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>13.0</td>
    </tr>
  </tbody>
</table>
<p>101 rows × 4 columns</p>
</div></div></div>
</div>
<section id="locate-any-tiles-not-just-perfect-matches">
<h4>Locate any tiles, not just perfect matches<a class="headerlink" href="#locate-any-tiles-not-just-perfect-matches" title="Permalink to this heading"></a></h4>
<p>Simply do not include the design file. A short (less than 120 bp) tile will probably appear with a significant number of reads. This is <strong>probably</strong> your untransformed background (cells transformed with a plasmid that did not get a tile). Most of the non-perfect matching tiles will be sequencing errors. You might find a way to map these back to their true tile. A few of the non-perfect matching tiles (likely those with a significant number of reads) will be sequencing library PCR amplification errors. An even fewer number of the non-perfect matching reads could be mutants that arose within the actual cell, or free tiles. Someone may put in the effort to distinguish these categories in the future.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">my_sort</span> <span class="o">=</span> <span class="n">sort</span><span class="o">.</span><span class="n">Sort</span><span class="p">([</span><span class="s2">&quot;../exampledata/bin1.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;../exampledata/bin2.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;../exampledata/bin3.fastq&quot;</span><span class="p">,</span> 
                <span class="s2">&quot;../exampledata/bin4.fastq&quot;</span><span class="p">],</span> <span class="n">bin_counts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100000</span><span class="p">,</span><span class="mi">100000</span><span class="p">,</span><span class="mi">100000</span><span class="p">,</span><span class="mi">100000</span><span class="p">],</span> 
               <span class="n">bin_values</span> <span class="o">=</span> <span class="p">[</span><span class="mi">61</span><span class="p">,</span><span class="mi">141</span><span class="p">,</span><span class="mi">251</span><span class="p">,</span><span class="mi">1462</span><span class="p">])</span>
<span class="n">activities_no_design</span><span class="p">,</span> <span class="n">numreads_total</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">my_sort</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">activities_no_design</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>Activity</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTAAATTCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT</th>
      <td>0.269721</td>
      <td>0.290977</td>
      <td>0.439302</td>
      <td>0.000000</td>
      <td>167.745576</td>
    </tr>
    <tr>
      <th>GAAGCTTTGTCTTTGTGTGATGATTTGTTGGGTGATCAAGATAGAGAAGAGAATGATAATGATGGTGATTTGAAAGATAAACAACCATGTATTGCTGATTATGCTCATTTGGGTCCAGAA</th>
      <td>0.106476</td>
      <td>0.534574</td>
      <td>0.333835</td>
      <td>0.025115</td>
      <td>202.380348</td>
    </tr>
    <tr>
      <th>GATTTGGCTGAAGATGATGAAGTTATGTGTATGGAAGATGAAGTTCAATCTATTCAACCAAATCATGAAAGACCAGATGATGGTCCAGAATTGGAACATGGTTTGGAGAATGGTGCTAAA</th>
      <td>0.368831</td>
      <td>0.495122</td>
      <td>0.136047</td>
      <td>0.000000</td>
      <td>126.458644</td>
    </tr>
    <tr>
      <th>GGTTAATTAAGGCGCGCCACTTCTAAATAAGCGA</th>
      <td>0.224803</td>
      <td>0.533542</td>
      <td>0.241655</td>
      <td>0.000000</td>
      <td>149.597758</td>
    </tr>
    <tr>
      <th>TCTGAATTGACTTCTACTTTGGGTATTTCTCATAGATTGCCACAATCTTTGACTCCATGTGTTAAGACTGGTTCTTTGCAATCTGGTGGTTTGGTTCAATCTGTTCCATTTGAAGAATTG</th>
      <td>0.351449</td>
      <td>0.278970</td>
      <td>0.369581</td>
      <td>0.000000</td>
      <td>153.537999</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCTACTGATTCTACTCCAATGTTTGATTATGATAATTTGGAAGATAATTCTAAAGATTGGACTTCTTTGTTTGATAATGATATTCCAGTTACTACTGATGATGTTTCTTTGGCTGATAAA</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>TCTACTGGTCAAGTCTTGTTTGATATTGATGACTTTAGATGGTTGTTGGATCCAGATGATGAACAATTGGGTAAAGAAGCTATCTTGTCTGATCAATTTGGTAAACCAACTCCAGAGAAT</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>AATACTCCAACTCCACCATCTTTGGTTGATGGTGTTGCTGGTGATGAAGAAGCATTTGATGAGATGTTTGATCCATTCTTTGAAGAATTGGATTCTATTCCAGAAGCTGCTTTGTGATAA</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>GAAGATCCAACTTCTGATTCTGCTATTCAACAATTGTGGAATCAAGGATTCTTGTTTGTTGAATCTTTGTCTTTGTGTGATGATTTGTTGGGTTCTCAAGATAGAGAAGAGAATACTAAT</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>GAAATTGATCAAATTTCTGATCCAGATAAATTGCCAGTTAATTTGGAACCATTTAGATTGGATCAATTGGAATTTACTGGTGATGATACTTCTGGTGCTGGTTTGAAATTTCAATGGGAT</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
  </tbody>
</table>
<p>112 rows × 5 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">numreads_total</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTAAATTCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT    44.0
GAAGCTTTGTCTTTGTGTGATGATTTGTTGGGTGATCAAGATAGAGAAGAGAATGATAATGATGGTGATTTGAAAGATAAACAACCATGTATTGCTGATTATGCTCATTTGGGTCCAGAA    21.0
GATTTGGCTGAAGATGATGAAGTTATGTGTATGGAAGATGAAGTTCAATCTATTCAACCAAATCATGAAAGACCAGATGATGGTCCAGAATTGGAACATGGTTTGGAGAATGGTGCTAAA    49.0
GGTTAATTAAGGCGCGCCACTTCTAAATAAGCGA                                                                                          24.0
TCTGAATTGACTTCTACTTTGGGTATTTCTCATAGATTGCCACAATCTTTGACTCCATGTGTTAAGACTGGTTCTTTGCAATCTGGTGGTTTGGTTCAATCTGTTCCATTTGAAGAATTG    70.0
                                                                                                                            ... 
TCTACTGATTCTACTCCAATGTTTGATTATGATAATTTGGAAGATAATTCTAAAGATTGGACTTCTTTGTTTGATAATGATATTCCAGTTACTACTGATGATGTTTCTTTGGCTGATAAA    24.0
TCTACTGGTCAAGTCTTGTTTGATATTGATGACTTTAGATGGTTGTTGGATCCAGATGATGAACAATTGGGTAAAGAAGCTATCTTGTCTGATCAATTTGGTAAACCAACTCCAGAGAAT    12.0
AATACTCCAACTCCACCATCTTTGGTTGATGGTGTTGCTGGTGATGAAGAAGCATTTGATGAGATGTTTGATCCATTCTTTGAAGAATTGGATTCTATTCCAGAAGCTGCTTTGTGATAA    14.0
GAAGATCCAACTTCTGATTCTGCTATTCAACAATTGTGGAATCAAGGATTCTTGTTTGTTGAATCTTTGTCTTTGTGTGATGATTTGTTGGGTTCTCAAGATAGAGAAGAGAATACTAAT    11.0
GAAATTGATCAAATTTCTGATCCAGATAAATTGCCAGTTAATTTGGAACCATTTAGATTGGATCAATTGGAATTTACTGGTGATGATACTTCTGGTGCTGGTTTGAAATTTCAATGGGAT    13.0
Length: 112, dtype: float64
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="get-data-for-reads-that-include-barcodes-and-tiles">
<h3>Get data for reads that include barcodes AND tiles<a class="headerlink" href="#get-data-for-reads-that-include-barcodes-and-tiles" title="Permalink to this heading"></a></h3>
<p>Support for barcoded only data may or may not be added in the future.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">activities_barcoded</span><span class="p">,</span> <span class="n">total_reads</span><span class="p">,</span> <span class="n">reads_per_bin</span> <span class="o">=</span> <span class="n">my_sort</span><span class="o">.</span><span class="n">process</span><span class="p">(</span><span class="n">barcoded</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Note that you get back less tiles using this method. The reason for this is because the read must have a locatable barcode AND tile, which is less likely than having one or the other. The primary reason for doing this analysis is assessing per-transformant variation. The assumption is that unique tile-barcode pairs come from unique original transformants.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">activities_barcoded</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>Activity</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTAAATTCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT</th>
      <th>ATCGTTACGTC</th>
      <td>0.289081</td>
      <td>0.312643</td>
      <td>0.398276</td>
      <td>0.000000</td>
      <td>161.683873</td>
    </tr>
    <tr>
      <th>GATTTGGCTGAAGATGATGAAGTTATGTGTATGGAAGATGAAGTTCAATCTATTCAACCAAATCATGAAAGACCAGATGATGGTCCAGAATTGGAACATGGTTTGGAGAATGGTGCTAAA</th>
      <th>GGAAGGTAGTA</th>
      <td>0.234149</td>
      <td>0.557113</td>
      <td>0.208738</td>
      <td>0.000000</td>
      <td>145.229204</td>
    </tr>
    <tr>
      <th>TCTGAATTGACTTCTACTTTGGGTATTTCTCATAGATTGCCACAATCTTTGACTCCATGTGTTAAGACTGGTTCTTTGCAATCTGGTGGTTTGGTTCAATCTGTTCCATTTGAAGAATTG</th>
      <th>GAGACCAAATC</th>
      <td>0.331376</td>
      <td>0.288808</td>
      <td>0.379816</td>
      <td>0.000000</td>
      <td>156.269710</td>
    </tr>
    <tr>
      <th>GATGCTAAGAGAGATTTGGAAGAATGTCAGAAGATTGTTGCTGATCCATCTAATATTGAATTGGATGCTCCACCAGAAGCTAGATTGGCTCAAGCTGAATTTGGTTCTCAAGATTCTGCT</th>
      <th>CAGGAAAATCG</th>
      <td>0.305253</td>
      <td>0.558685</td>
      <td>0.136062</td>
      <td>0.000000</td>
      <td>131.546634</td>
    </tr>
    <tr>
      <th>CAAGACTTTATTTGTTCTGTTACTCAAGATGTTGGTATTAATTATACTCATCCACAGAATTTGCCAGGTGTTTCTAAAGATGGTACTTCTGTCTTCTTCTTTAATAAGACTGCTCATGCT</th>
      <th>ACATGAACTTA</th>
      <td>0.244871</td>
      <td>0.416160</td>
      <td>0.280666</td>
      <td>0.058303</td>
      <td>229.301369</td>
    </tr>
    <tr>
      <th>...</th>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>AATACTCCAACTCCACCATCTTTGGTTGATGGTGTTGCTGGTGATGAAGAAGCATTTGATGATTTGAAGATGTTTGATCCATTCTTTGAAGAATTGGATTCTATTCCAGAAGCTGCTTTG</th>
      <th>CTGTCTTCATA</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>AAAGATAGAGATAATGAGAATCAAGGTGAAGAAGATCCAACTTGGTTTATTGATGCTTTGCAATTTAATGGTTTGGTTCAAGCTGAATCTGCTCAATTGTCTTTGTGTGATGATTTGTTG</th>
      <th>GTGGCAGTATG</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>TATCCACCAAGAGATGATTATCATGAACAACCATATCCAGTTATGGATACTGAAGGTAATGTTATTGGTTTGACTGATTTGGAATGTGATGAAGAATTGTTGCCAATTTCTTGGGATTGG</th>
      <th>CACCCACTCCG</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>TCTACTGATTCTACTCCAATGTTTGATTATGATAATTTGGAAGATAATTCTAAAGATTGGACTTCTTTGTTTGATAATGATATTCCAGTTACTACTGATGATGTTTCTTTGGCTGATAAA</th>
      <th>CTCAGTAATGG</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
    <tr>
      <th>AATACTCCAACTCCACCATCTTTGGTTGATGGTGTTGCTGGTGATGAAGAAGCATTTGATGAGATGTTTGATCCATTCTTTGAAGAATTGGATTCTATTCCAGAAGCTGCTTTGTGATAA</th>
      <th>CCAGCAAGAAT</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1462.000000</td>
    </tr>
  </tbody>
</table>
<p>77 rows × 5 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">format_index</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
    <span class="k">return</span> <span class="s2">&quot;font-size: 6pt;&quot;</span>
<span class="n">activities_no_design</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">applymap_index</span><span class="p">(</span><span class="n">format_index</span><span class="p">)</span>
<span class="n">activities_no_design</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">apply</span><span class="p">({</span><span class="s2">&quot;font-size&quot;</span><span class="p">:</span><span class="s2">&quot;6pt&quot;</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">KeyError</span><span class="g g-Whitespace">                                  </span>Traceback (most recent call last)
<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/IPython/core/formatters.py:342,</span> in <span class="ni">BaseFormatter.__call__</span><span class="nt">(self, obj)</span>
<span class="g g-Whitespace">    </span><span class="mi">340</span>     <span class="n">method</span> <span class="o">=</span> <span class="n">get_real_method</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">print_method</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">341</span>     <span class="k">if</span> <span class="n">method</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">342</span>         <span class="k">return</span> <span class="n">method</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">343</span>     <span class="k">return</span> <span class="kc">None</span>
<span class="g g-Whitespace">    </span><span class="mi">344</span> <span class="k">else</span><span class="p">:</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/io/formats/style.py:385,</span> in <span class="ni">Styler._repr_html_</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">380</span><span class="w"> </span><span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">381</span><span class="sd"> Hooks into Jupyter notebook rich display system, which calls _repr_html_ by</span>
<span class="g g-Whitespace">    </span><span class="mi">382</span><span class="sd"> default if an object is returned at the end of a cell.</span>
<span class="g g-Whitespace">    </span><span class="mi">383</span><span class="sd"> &quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">384</span> <span class="k">if</span> <span class="n">get_option</span><span class="p">(</span><span class="s2">&quot;styler.render.repr&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;html&quot;</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">385</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_html</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">386</span> <span class="k">return</span> <span class="kc">None</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/io/formats/style.py:1377,</span> in <span class="ni">Styler.to_html</span><span class="nt">(self, buf, table_uuid, table_attributes, sparse_index, sparse_columns, bold_headers, caption, max_rows, max_columns, encoding, doctype_html, exclude_styles, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">1374</span>     <span class="n">obj</span><span class="o">.</span><span class="n">set_caption</span><span class="p">(</span><span class="n">caption</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1376</span> <span class="c1"># Build HTML string..</span>
<span class="ne">-&gt; </span><span class="mi">1377</span> <span class="n">html</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">_render_html</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1378</span>     <span class="n">sparse_index</span><span class="o">=</span><span class="n">sparse_index</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1379</span>     <span class="n">sparse_columns</span><span class="o">=</span><span class="n">sparse_columns</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1380</span>     <span class="n">max_rows</span><span class="o">=</span><span class="n">max_rows</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1381</span>     <span class="n">max_cols</span><span class="o">=</span><span class="n">max_columns</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1382</span>     <span class="n">exclude_styles</span><span class="o">=</span><span class="n">exclude_styles</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1383</span>     <span class="n">encoding</span><span class="o">=</span><span class="n">encoding</span> <span class="ow">or</span> <span class="n">get_option</span><span class="p">(</span><span class="s2">&quot;styler.render.encoding&quot;</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1384</span>     <span class="n">doctype_html</span><span class="o">=</span><span class="n">doctype_html</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1385</span>     <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1386</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1388</span> <span class="k">return</span> <span class="n">save_to_buffer</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1389</span>     <span class="n">html</span><span class="p">,</span> <span class="n">buf</span><span class="o">=</span><span class="n">buf</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="p">(</span><span class="n">encoding</span> <span class="k">if</span> <span class="n">buf</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1390</span> <span class="p">)</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/io/formats/style_render.py:206,</span> in <span class="ni">StylerRenderer._render_html</span><span class="nt">(self, sparse_index, sparse_columns, max_rows, max_cols, **kwargs)</span>
<span class="g g-Whitespace">    </span><span class="mi">194</span> <span class="k">def</span> <span class="nf">_render_html</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">195</span>     <span class="bp">self</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">196</span>     <span class="n">sparse_index</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">200</span>     <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">201</span> <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">202</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">203</span><span class="sd">     Renders the ``Styler`` including all applied styles to HTML.</span>
<span class="g g-Whitespace">    </span><span class="mi">204</span><span class="sd">     Generates a dict with necessary kwargs passed to jinja2 template.</span>
<span class="g g-Whitespace">    </span><span class="mi">205</span><span class="sd">     &quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">206</span>     <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_render</span><span class="p">(</span><span class="n">sparse_index</span><span class="p">,</span> <span class="n">sparse_columns</span><span class="p">,</span> <span class="n">max_rows</span><span class="p">,</span> <span class="n">max_cols</span><span class="p">,</span> <span class="s2">&quot;&amp;nbsp;&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">207</span>     <span class="n">d</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">208</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">template_html</span><span class="o">.</span><span class="n">render</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">209</span>         <span class="o">**</span><span class="n">d</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">210</span>         <span class="n">html_table_tpl</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">template_html_table</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">211</span>         <span class="n">html_style_tpl</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">template_html_style</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">212</span>     <span class="p">)</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/io/formats/style_render.py:163,</span> in <span class="ni">StylerRenderer._render</span><span class="nt">(self, sparse_index, sparse_columns, max_rows, max_cols, blank)</span>
<span class="g g-Whitespace">    </span><span class="mi">149</span> <span class="k">def</span> <span class="nf">_render</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">150</span>     <span class="bp">self</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">151</span>     <span class="n">sparse_index</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">155</span>     <span class="n">blank</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">156</span> <span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">157</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">158</span><span class="sd">     Computes and applies styles and then generates the general render dicts.</span>
<span class="g g-Whitespace">    </span><span class="mi">159</span><span class="sd"> </span>
<span class="g g-Whitespace">    </span><span class="mi">160</span><span class="sd">     Also extends the `ctx` and `ctx_index` attributes with those of concatenated</span>
<span class="g g-Whitespace">    </span><span class="mi">161</span><span class="sd">     stylers for use within `_translate_latex`</span>
<span class="g g-Whitespace">    </span><span class="mi">162</span><span class="sd">     &quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">163</span>     <span class="bp">self</span><span class="o">.</span><span class="n">_compute</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">164</span>     <span class="n">dxs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="g g-Whitespace">    </span><span class="mi">165</span>     <span class="n">ctx_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/io/formats/style_render.py:258,</span> in <span class="ni">StylerRenderer._compute</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">256</span> <span class="n">r</span> <span class="o">=</span> <span class="bp">self</span>
<span class="g g-Whitespace">    </span><span class="mi">257</span> <span class="k">for</span> <span class="n">func</span><span class="p">,</span> <span class="n">args</span><span class="p">,</span> <span class="n">kwargs</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_todo</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">258</span>     <span class="n">r</span> <span class="o">=</span> <span class="n">func</span><span class="p">(</span><span class="bp">self</span><span class="p">)(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">259</span> <span class="k">return</span> <span class="n">r</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/io/formats/style.py:1736,</span> in <span class="ni">Styler._apply</span><span class="nt">(self, func, axis, subset, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">1734</span> <span class="n">axis</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_get_axis_number</span><span class="p">(</span><span class="n">axis</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1735</span> <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1736</span>     <span class="n">result</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1737</span> <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1738</span>     <span class="n">result</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># see GH 42005</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/frame.py:9568,</span> in <span class="ni">DataFrame.apply</span><span class="nt">(self, func, axis, raw, result_type, args, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">9557</span> <span class="kn">from</span> <span class="nn">pandas.core.apply</span> <span class="kn">import</span> <span class="n">frame_apply</span>
<span class="g g-Whitespace">   </span><span class="mi">9559</span> <span class="n">op</span> <span class="o">=</span> <span class="n">frame_apply</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">9560</span>     <span class="bp">self</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">9561</span>     <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">9566</span>     <span class="n">kwargs</span><span class="o">=</span><span class="n">kwargs</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">9567</span> <span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">9568</span> <span class="k">return</span> <span class="n">op</span><span class="o">.</span><span class="n">apply</span><span class="p">()</span><span class="o">.</span><span class="n">__finalize__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;apply&quot;</span><span class="p">)</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/apply.py:735,</span> in <span class="ni">FrameApply.apply</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">733</span> <span class="c1"># dispatch to agg</span>
<span class="g g-Whitespace">    </span><span class="mi">734</span> <span class="k">if</span> <span class="n">is_list_like</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">f</span><span class="p">):</span>
<span class="ne">--&gt; </span><span class="mi">735</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply_multiple</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">737</span> <span class="c1"># all empty</span>
<span class="g g-Whitespace">    </span><span class="mi">738</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/apply.py:591,</span> in <span class="ni">Apply.apply_multiple</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">582</span> <span class="k">def</span> <span class="nf">apply_multiple</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span> <span class="o">|</span> <span class="n">Series</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">583</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">584</span><span class="sd">     Compute apply in case of a list-like or dict-like.</span>
<span class="g g-Whitespace">    </span><span class="mi">585</span><span class="sd"> </span>
<span class="sd">   (...)</span>
<span class="g g-Whitespace">    </span><span class="mi">589</span><span class="sd">         Result when self.f is a list-like or dict-like, None otherwise.</span>
<span class="g g-Whitespace">    </span><span class="mi">590</span><span class="sd">     &quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">591</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">obj</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">axis</span><span class="p">,</span> <span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span><span class="p">)</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/frame.py:9342,</span> in <span class="ni">DataFrame.aggregate</span><span class="nt">(self, func, axis, *args, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">9339</span> <span class="n">relabeling</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">order</span> <span class="o">=</span> <span class="n">reconstruct_func</span><span class="p">(</span><span class="n">func</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">9341</span> <span class="n">op</span> <span class="o">=</span> <span class="n">frame_apply</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span> <span class="n">kwargs</span><span class="o">=</span><span class="n">kwargs</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">9342</span> <span class="n">result</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">agg</span><span class="p">()</span>
<span class="g g-Whitespace">   </span><span class="mi">9344</span> <span class="k">if</span> <span class="n">relabeling</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">9345</span>     <span class="c1"># This is to keep the order to columns occurrence unchanged, and also</span>
<span class="g g-Whitespace">   </span><span class="mi">9346</span>     <span class="c1"># keep the order of new columns occurrence unchanged</span>
<span class="g g-Whitespace">   </span><span class="mi">9347</span> 
<span class="g g-Whitespace">   </span><span class="mi">9348</span>     <span class="c1"># For the return values of reconstruct_func, if relabeling is</span>
<span class="g g-Whitespace">   </span><span class="mi">9349</span>     <span class="c1"># False, columns and order will be None.</span>
<span class="g g-Whitespace">   </span><span class="mi">9350</span>     <span class="k">assert</span> <span class="n">columns</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/apply.py:776,</span> in <span class="ni">FrameApply.agg</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">774</span> <span class="n">result</span> <span class="o">=</span> <span class="kc">None</span>
<span class="g g-Whitespace">    </span><span class="mi">775</span> <span class="k">try</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">776</span>     <span class="n">result</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">agg</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">777</span> <span class="k">except</span> <span class="ne">TypeError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">778</span>     <span class="n">exc</span> <span class="o">=</span> <span class="ne">TypeError</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">779</span>         <span class="s2">&quot;DataFrame constructor called with &quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">780</span>         <span class="sa">f</span><span class="s2">&quot;incompatible data and dtype: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">781</span>     <span class="p">)</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/apply.py:172,</span> in <span class="ni">Apply.agg</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">169</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply_str</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">171</span> <span class="k">if</span> <span class="n">is_dict_like</span><span class="p">(</span><span class="n">arg</span><span class="p">):</span>
<span class="ne">--&gt; </span><span class="mi">172</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">agg_dict_like</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">173</span> <span class="k">elif</span> <span class="n">is_list_like</span><span class="p">(</span><span class="n">arg</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">174</span>     <span class="c1"># we require a list, but not a &#39;str&#39;</span>
<span class="g g-Whitespace">    </span><span class="mi">175</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">agg_list_like</span><span class="p">()</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/apply.py:496,</span> in <span class="ni">Apply.agg_dict_like</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">493</span>     <span class="n">selected_obj</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">_selected_obj</span>
<span class="g g-Whitespace">    </span><span class="mi">494</span>     <span class="n">selection</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">_selection</span>
<span class="ne">--&gt; </span><span class="mi">496</span> <span class="n">arg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_dictlike_arg</span><span class="p">(</span><span class="s2">&quot;agg&quot;</span><span class="p">,</span> <span class="n">selected_obj</span><span class="p">,</span> <span class="n">arg</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">498</span> <span class="k">if</span> <span class="n">selected_obj</span><span class="o">.</span><span class="n">ndim</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">499</span>     <span class="c1"># key only used for output</span>
<span class="g g-Whitespace">    </span><span class="mi">500</span>     <span class="n">colg</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">_gotitem</span><span class="p">(</span><span class="n">selection</span><span class="p">,</span> <span class="n">ndim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="nn">File ~/bioinformatics_exploration/.venv/lib/python3.8/site-packages/pandas/core/apply.py:619,</span> in <span class="ni">Apply.normalize_dictlike_arg</span><span class="nt">(self, how, obj, func)</span>
<span class="g g-Whitespace">    </span><span class="mi">617</span>     <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">618</span>         <span class="n">cols_sorted</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">safe_sort</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">cols</span><span class="p">)))</span>
<span class="ne">--&gt; </span><span class="mi">619</span>         <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column(s) </span><span class="si">{</span><span class="n">cols_sorted</span><span class="si">}</span><span class="s2"> do not exist&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">621</span> <span class="n">aggregator_types</span> <span class="o">=</span> <span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">623</span> <span class="c1"># if we have a dict of any non-scalars</span>
<span class="g g-Whitespace">    </span><span class="mi">624</span> <span class="c1"># eg. {&#39;A&#39; : [&#39;mean&#39;]}, normalize all to</span>
<span class="g g-Whitespace">    </span><span class="mi">625</span> <span class="c1"># be list-likes</span>
<span class="g g-Whitespace">    </span><span class="mi">626</span> <span class="c1"># Cannot use func.values() because arg may be a Series</span>

<span class="ne">KeyError</span>: &quot;Column(s) [&#39;font-size&#39;] do not exist&quot;
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;pandas.io.formats.style.Styler at 0x7ff314bdbdf0&gt;
</pre></div>
</div>
</div>
</div>
</section>
<section id="support-for-custom-anchor-sequences">
<h3>Support for custom anchor sequences<a class="headerlink" href="#support-for-custom-anchor-sequences" title="Permalink to this heading"></a></h3>
<p>If your tiles or barcodes have a custom anchor sequence (AKA the non-variable portion of the read that is used to locate the variable portion of the read), you can specify that in the kwargs of your Sort(). This passes the arguments to the pull_A() function used on each read to locate the sequence of interest (AKA AD or tile).</p>
<section id="use-the-sequence-directly-preceding-the-ad-tile-as-an-anchor-sequence-additional-characters-between-ad-tile-and-preceding-anchor-seqeunce-will-not-work">
<h4>Use the sequence directly preceding the AD/tile as an anchor sequence. Additional characters between AD/tile and preceding anchor seqeunce will not work<a class="headerlink" href="#use-the-sequence-directly-preceding-the-ad-tile-as-an-anchor-sequence-additional-characters-between-ad-tile-and-preceding-anchor-seqeunce-will-not-work" title="Permalink to this heading"></a></h4>
<p>My preceding anchor sequence is in blue while my AD/tile sequence is in green in this example read.</p>
<p>The anchor sequence preceding the barcode is purple while the anchor sequence anteceding is red. Barcode length is 11 by default. (Not necessary if you don’t have barcodes)</p>
<section id="example-read">
<h5>Example read<a class="headerlink" href="#example-read" title="Permalink to this heading"></a></h5>
<p><span style="font-size:10px">TCCCTGCGGGCTCTACTTCATCG<span style="color:blue;font-weight:800">GCTAGC</span><span style="color:green;font-weight:500;font-size:10px"><strong>GGTTCTTCTAAATTGAGATGTGATAATAATGCTGCTGCTCATGTTAAATTGGATTCATTTCCAGCTGGTGTTAGATTTGATACATCTGATGAAGAATTGTTGGAACATTTGGCTGCTAAA</strong></span>TGATAAATAGATGA<span style="color:purple;font-weight:800">GGGCCCG</span>TCAACATAGAA<span style="color:red;font-weight:800">GGAGAGAA</span>ACATCTAAAAAAGCGATA</span></p>
</section>
</section>
<section id="specify-alternate-values-in-a-dictionary-and-pass-that-dictionary-to-sort-process">
<h4>Specify alternate values in a dictionary and pass that dictionary to Sort.process()<a class="headerlink" href="#specify-alternate-values-in-a-dictionary-and-pass-that-dictionary-to-sort-process" title="Permalink to this heading"></a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># these are the default values which would work for my example read above</span>
<span class="c1"># no input is required if they work for you</span>
<span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ad_preceder&quot;</span><span class="p">:</span><span class="s2">&quot;GCTAGC&quot;</span><span class="p">,</span> <span class="s2">&quot;bc_preceder&quot;</span><span class="p">:</span><span class="s2">&quot;GGGCCCG&quot;</span><span class="p">,</span> <span class="s2">&quot;bc_anteceder&quot;</span><span class="p">:</span><span class="s2">&quot;GGAGAGAA&quot;</span><span class="p">,</span> <span class="s2">&quot;bclength&quot;</span><span class="p">:</span><span class="mi">11</span><span class="p">,</span> <span class="s2">&quot;ad_length&quot;</span><span class="p">:</span><span class="mi">120</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">activities_barcoded</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">my_sort</span><span class="o">.</span><span class="n">process</span><span class="p">(</span><span class="n">barcoded</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_colwidth&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#activities_barcoded.style.set_properties(**{&#39;background-color&#39;: &#39;grey&#39;,&#39;font-size&#39;: &#39;6pt&#39;})</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="example-output-for-pull-ad">
<h4>Example output for pull_AD<a class="headerlink" href="#example-output-for-pull-ad" title="Permalink to this heading"></a></h4>
<p>This might be useful for someone who wants to use pull_AD to analyze reads for a purpose outside of calculating activities.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">labtools.adtools.finder</span> <span class="kn">import</span> <span class="n">pull_AD</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">read</span> <span class="o">=</span> <span class="s2">&quot;TCCCTGCGGGCTCTACTTCATCGGCTAGCGGTTCTTCTAAATTGAGATGTGATAATAATGCTGCTGCTCATGTTAAATTGGATTCATTTCCAGCTGGTGTTAGATTTGATACATCTGATGAAGAATTGTTGGAACATTTGGCTGCTAAATGATAAATAGATGAGGGCCCGTCAACATAGAAGGAGAGAAACATCTAAAAAAGCGATA&quot;</span>
<span class="n">pull_AD</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;GGTTCTTCTAAATTGAGATGTGATAATAATGCTGCTGCTCATGTTAAATTGGATTCATTTCCAGCTGGTGTTAGATTTGATACATCTGATGAAGAATTGTTGGAACATTTGGCTGCTAAA&#39;,
 &#39;TCAACATAGAA&#39;)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="counting-tiles-in-a-fastq-file">
<h3>Counting tiles in a fastq file<a class="headerlink" href="#counting-tiles-in-a-fastq-file" title="Permalink to this heading"></a></h3>
<p>This essentially applies the pull_AD function shown above to every read in the fastq file. You can also pass the kwargs dict to it to specify custom anchor sequences.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">labtools.adtools.pair_counter</span> <span class="kn">import</span> <span class="n">seq_counter</span>
<span class="n">seq_counter</span><span class="p">(</span><span class="s2">&quot;../exampledata/mini.fastq&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GGTTCTTCTAAATTGAGATGTGATAATAATGCTGCTGCTCATGTTAAATTGGATTCATTTCCAGCTGGTGTTAGATTTGATACATCTGATGAAGAATTGTTGGAACATTTGGCTGCTAAA    1
GAAGAATTGTTTTTACATTTGTCTGCTAAGATTGGTAGATCTTCTAGGAAACCACATCCATTCTTGGATGAATTTATTCATACTTTGGTTGAAGAAGATGGTATTTGTAGAACTCATCCA    3
dtype: int64
</pre></div>
</div>
</div>
</div>
<section id="use-barcoded-true-to-count-tile-and-barcode-pairs">
<h4>Use barcoded = True  to count tile and barcode pairs<a class="headerlink" href="#use-barcoded-true-to-count-tile-and-barcode-pairs" title="Permalink to this heading"></a></h4>
</section>
<section id="use-the-design-file-flag-to-look-for-specific-tiles">
<h4>Use the design_file flag to look for specific tiles<a class="headerlink" href="#use-the-design-file-flag-to-look-for-specific-tiles" title="Permalink to this heading"></a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq_counter</span><span class="p">(</span><span class="s2">&quot;../exampledata/bin1.fastq&quot;</span><span class="p">,</span> <span class="n">barcoded</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">design_to_use</span><span class="o">=</span><span class="s2">&quot;../exampledata/unique_seqs.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTAAATTCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT  ATCGTTACGTC    11.0
GAAGCTTTGTCTTTGTGTGATGATTTGTTGGGTGATCAAGATAGAGAAGAGAATGATAATGATGGTGATTTGAAAGATAAACAACCATGTATTGCTGATTATGCTCATTTGGGTCCAGAA  ACCATTTAGCG    1.0 
GATTTGGCTGAAGATGATGAAGTTATGTGTATGGAAGATGAAGTTCAATCTATTCAACCAAATCATGAAAGACCAGATGATGGTCCAGAATTGGAACATGGTTTGGAGAATGGTGCTAAA  GGAAGGTAGTA    5.0 
GGTCAAAGGAAGAGAAGGAAGATTACTCCAACTTTGGTTAATGATGAACCAGTTAGATGGCATAAGACTGGTAGAACTAAACCAGTTATGTTGTCTGGTGTTCAAAGAGGTTGTAAGAAA  TAGCTCAAACC    4.0 
TCTGAATTGACTTCTACTTTGGGTATTTCTCATAGATTGCCACAATCTTTGACTCCATGTGTTAAGACTGGTTCTTTGCAATCTGGTGGTTTGGTTCAATCTGTTCCATTTGAAGAATTG  GAGACCAAATC    21.0
                                                                                                                                         ... 
CAAGCTATTGCTAATGATCAACAAGCTGTTGAAGCTGCTTCTGCTTGTGATGATGCTGCTGGTTCTCAAGATAGAGAAGAGAATACTAATTCTGGTTCTGCTAAAGATAAACAACCATGT  TCGTCTGACAC    1.0 
AAAGATAGAGATAATGAGAATCAAGGTGAAGAAGATCCAACTTGGTTTGATTCTGGTTCTCAATTTATCTTGAATGCTCAACAATTGGTTGAAGCTTTGTCTTTGTGTGATGATTTGTTG  TTCTTTCCCTA    1.0 
CATTTGGGTATTGAAGAAGATGAGAAAGAAGGTGATGAAGTTGTTTCTAAGATCTTCTATCAACAACCACAACAATTGGTTGTTAAGAGAGGTGATAAAGCTGAACAAGAAGTTGATGAA  ACTCCCCGCAA    1.0 
GAAGATCCAACTGCATTTGATTCTGGTGATCAAGCTATCTTGAATGATCAACAAGCTGTTGAAGCTTTGTCTGCTTGTGATGATTTGGCTGGTTCTCAAGATAGAGAAGAGAATACTAAT  GGTTGAAGTGC    1.0 
GAAGAAACTGTTGTCTTGCCACCATCTGTTAAGACTACTAAAGATGGTGATAATTTGCAATCTCAAGATCAGAAATTGTGGGAAGGTGATTCTCAATTTGAATTGTTGGATTCTCAACAA  ACCACCGACGA    1.0 
Length: 684, dtype: float64
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="labtools" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="changelog.html" class="btn btn-neutral float-right" title="Changelog" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Jordan Stefani.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>